sys/trickle/trickle.c

```diff
@@ -30,63 +27,61 @@ void trickle_callback(trickle_t *trickle)
     if ((trickle->c < trickle->k) || (trickle->k == 0)) {
         (*trickle->callback.func)(trickle->callback.args);
     }
+
+    trickle_interval(trickle);
 }
 
 void trickle_interval(trickle_t *trickle)
 {
-    uint32_t max_interval;
+    uint32_t old_interval = trickle->I;
+    uint32_t max_interval = trickle->Imin << trickle->Imax;
+    uint32_t diff = old_interval - trickle->t;
 
-    trickle->I = trickle->I * 2;
-    max_interval = trickle->Imin << trickle->Imax;
+    trickle->I *= 2;
 
     if ((trickle->I == 0) || (trickle->I > max_interval)) {
         trickle->I = max_interval;
     }
 
-    DEBUG("trickle: I == %" PRIu32 "\n", trickle->I);
+    DEBUG("trickle: I == %" PRIu32 ", diff == %" PRIu32 "\n", trickle->I, diff);
 
     trickle->c = 0;
-    trickle->t = (trickle->I / 2) + random_uint32_range(0, (trickle->I / 2) + 1);
+    trickle->t = random_uint32_range(old_interval, trickle->I);
 
-    trickle->msg_callback_time = trickle->t * MS_PER_SEC;
-    xtimer_set_msg64(&trickle->msg_callback_timer, trickle->msg_callback_time,
-                     &trickle->msg_callback, trickle->pid);
+    trickle->msg_time = (trickle->t + diff) * MS_PER_SEC;
```

Comment:
yep, but that's intentional:

|----,----|--,-----|------,--|---...
I1   t1  I2  t2   I3      t3 I4

In the current implementation, timers are only thrown at tx (x=1..n).
I.e.: we are currently at t1 and we want to calculate t2 now (the value for xtimer_msg). For this, we move I1 to I2 and do the calculation as before, but now we have do att the dif (I2-t1) to t2 to account for optimization that removes the second msg_timer.